numba/tests/test_linalg.py

```diff
@@ -344,31 +349,181 @@ def sample_vector(self, n, dtype):
-        # with a kwarg for condition number?
-        break_at = 10
-        jmp = 0
-        # have a few attempts at shuffling to get the condition number down
-        # else not worry about it
-        mn = size[0] * size[1]
+
+        # default condition
+        d_cond = 1.
+
+        if len(size) != 2:
+            raise ValueError("size must be a length 2 tuple.")
+
+        if order not in ['F', 'C']:
+            raise ValueError("order must be one of 'F' or 'C'.")
+
+        if dtype not in [np.float32, np.float64, np.complex64, np.complex128]:
+            raise ValueError("dtype must be a numpy floating point type.")
+
+        if rank is not None and condition is not None:
+            raise ValueError("Only one of rank or condition can be specified.")
+
+        if condition is None:
+            condition = d_cond
+
+        if condition < 1:
+            raise ValueError("Condition number must be >=1.")
+
         np.random.seed(0)  # repeatable seed
-        while jmp < break_at:
-            v = self.sample_vector(mn, dtype)
-            # shuffle to improve conditioning
-            np.random.shuffle(v)
-            A = np.reshape(v, size)
-            if np.linalg.cond(A) < mn:
-                return np.array(A, order=order, dtype=dtype)
-            jmp += 1
-        return A
+        m, n = size
+
+        if m < 0 or n < 0:
+            raise ValueError("Negative dimensions given for matrix shape.")
+
+        minmn = min(m, n)
+        if rank is None:
+            rv = minmn
+        else:
+            if rank <= 0:
+                raise ValueError("Rank must be greater than zero.")
+            if not isinstance(rank, Integral):
+                raise ValueError("Rank must an integer.")
+            rv = rank
+            if rank > minmn:
+                raise ValueError("Rank given greater than full rank.")
+
+        # Build a sample matrix via combining SVD like inputs
+        # this ought to use MGS
+        tmp = self.sample_vector(m * m, dtype).reshape(m, m)
+        U, _ = np.linalg.qr(tmp)
+        tmp = self.sample_vector(n * n, dtype).reshape(n, n)
+        V, _ = np.linalg.qr(tmp)
+
+        if m == 1 or n == 1:
+            # vector, must be rank 1 (enforced above)
+            # condition of vector is also 1
+            if condition != d_cond:
+                raise ValueError(
+                    "Condition number was specified for a vector (always 1.).")
+            maxmn = max(m, n)
+            Q = self.sample_vector(maxmn, dtype).reshape(m, n)
+        else:
+            sv = np.linspace(d_cond, condition, rv)
+            S = np.zeros((m, n))
+            idx = np.nonzero(np.eye(m, n))
+            S[idx[0][:rv], idx[1][:rv]] = sv
+            Q = np.dot(np.dot(U, S), V.T)  # construct
+            Q = np.array(Q, dtype=dtype, order=order)  # sort out order/type
+
+        return Q
+
+    # the sample matrix code is a bit involved, so we'll test it!
+    def test_specific_sample_matrix(self):
```

Comment:
This test needs pulling out of here as this is a base class and it doesn't need to be inherited elsewhere.